GCP Sample App Fails

This happens even when upgrading to the latest version of the `function:run` plugin.
We probably need to add an integration test for the sample that runs `function:run` directly.

```
WARNING: Failed to execute org.springframework.cloud.function.adapter.gcloud.FunctionInvoker
java.lang.NoClassDefFoundError: org/json/JSONObject
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.isJson(SimpleFunctionRegistry.java:849)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.convertInputValueIfNecessary(SimpleFunctionRegistry.java:817)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.doApply(SimpleFunctionRegistry.java:625)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.apply(SimpleFunctionRegistry.java:479)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.apply(SimpleFunctionRegistry.java:467)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.invokeFunction(SimpleFunctionRegistry.java:535)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.doApply(SimpleFunctionRegistry.java:623)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.apply(SimpleFunctionRegistry.java:479)
        at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.apply(SimpleFunctionRegistry.java:467)
        at org.springframework.cloud.function.adapter.gcloud.FunctionInvoker.service(FunctionInvoker.java:93)
        at com.google.cloud.functions.invoker.NewHttpFunctionExecutor.service(NewHttpFunctionExecutor.java:65)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:547)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1297)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1212)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at com.google.cloud.functions.invoker.runner.Invoker$NotFoundHandler.handle(Invoker.java:397)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:500)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:270)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: org.json.JSONObject
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 35 more
```
cc @dzou @dmitry-s

--- a/spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/main/java/org/springframework/cloud/function/adapter/gcp/GcfJarLauncher.java
+++ b/spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/main/java/org/springframework/cloud/function/adapter/gcp/GcfJarLauncher.java
@@ -16,7 +16,8 @@
 
 package org.springframework.cloud.function.adapter.gcp;
 
-import java.util.List;
+import java.util.Collections;
+import java.util.Iterator;
 
 import com.google.cloud.functions.Context;
 import com.google.cloud.functions.HttpFunction;
@@ -27,7 +28,6 @@ import com.google.cloud.functions.RawBackgroundFunction;
 import org.springframework.boot.loader.JarLauncher;
 import org.springframework.boot.loader.archive.Archive;
 import org.springframework.boot.loader.jar.JarFile;
-import org.springframework.util.CollectionUtils;
 
 /**
  * The launcher class written at the top-level of the output JAR to be deployed to
@@ -65,12 +65,14 @@ public class GcfJarLauncher extends JarLauncher implements HttpFunction, RawBack
 		((RawBackgroundFunction) delegate).accept(json, context);
 	}
 
-	private List<Archive> discoverClassPathAcrhives() throws Exception {
-		List<Archive> classPathArchives = getClassPathArchives();
-		if (CollectionUtils.isEmpty(classPathArchives)) {
-			classPathArchives.add(this.getArchive());
+	private Iterator<Archive> discoverClassPathAcrhives() throws Exception {
+		Iterator<Archive> classPathArchives = getClassPathArchivesIterator();
+		if (classPathArchives.hasNext()) {
+			return classPathArchives;
 		}
-		return classPathArchives;
+		else {
+			return Collections.singleton(this.getArchive()).iterator();
+		}
 	}
 }
 

--- a/spring-cloud-function-samples/function-sample-gcp-background/pom.xml
+++ b/spring-cloud-function-samples/function-sample-gcp-background/pom.xml
@@ -19,11 +19,15 @@
 		<relativePath/> <!-- lookup parent from repository -->
 	</parent>
 
+	<properties>
+		<spring-cloud-function.version>3.0.8.BUILD-SNAPSHOT</spring-cloud-function.version>
+	</properties>
+
 	<dependencies>
 		<dependency>
 			<groupId>org.springframework.cloud</groupId>
 			<artifactId>spring-cloud-function-adapter-gcp</artifactId>
-			<version>3.1.0.BUILD-SNAPSHOT</version>
+			<version>${spring-cloud-function.version}</version>
 		</dependency>
 	</dependencies>
 
@@ -46,7 +50,7 @@
 					<dependency>
 						<groupId>org.springframework.cloud</groupId>
 						<artifactId>spring-cloud-function-adapter-gcp</artifactId>
-						<version>3.1.0.BUILD-SNAPSHOT</version>
+						<version>${spring-cloud-function.version}</version>
 					</dependency>
 				</dependencies>
 			</plugin>

--- a/spring-cloud-function-samples/function-sample-gcp-http/pom.xml
+++ b/spring-cloud-function-samples/function-sample-gcp-http/pom.xml
@@ -19,11 +19,15 @@
 		<relativePath/> <!-- lookup parent from repository -->
 	</parent>
 
+	<properties>
+		<spring-cloud-function.version>3.0.8.BUILD-SNAPSHOT</spring-cloud-function.version>
+	</properties>
+
 	<dependencies>
 		<dependency>
 			<groupId>org.springframework.cloud</groupId>
 			<artifactId>spring-cloud-function-adapter-gcp</artifactId>
-			<version>3.1.0.BUILD-SNAPSHOT</version>
+			<version>${spring-cloud-function.version}</version>
 		</dependency>
 
 		<!-- test dependencies -->
@@ -74,7 +78,7 @@
 					<dependency>
 						<groupId>org.springframework.cloud</groupId>
 						<artifactId>spring-cloud-function-adapter-gcp</artifactId>
-						<version>3.1.0.BUILD-SNAPSHOT</version>
+						<version>${spring-cloud-function.version}</version>
 					</dependency>
 				</dependencies>
 			</plugin>

